apps/labs/posts/python-wheels-from-tags-to-variants.md

```diff
@@ -0,0 +1,1146 @@
+list whenever they are about to use a new property value.
+However, plugins are scoped to themselves and can only order feature names
+and values. They cannot provide namespace ordering, as that would effectively
+mean one plugin deciding how important another plugin is.
+
+This leaves us with two possibilities: either the package author or the user
+needs to define namespace ordering. Originally, we went with the latter idea;
+after all, the users know best whether they prefer CUDA or ROCm, or perhaps CPU
+optimization. However, this meant that the user had to jump through
+the hoops of configuring variant usage first, and once again things could not
+work out of the box. So we went with the next best thing possible: requiring package
+maintainers to specify namespace ordering in the variant configuration; after all, package authors
+are in good position to specify which variants are the most beneficial to their software.
+
+<figure>
+
+```toml
+[variant.default-priorities]
+# prefer aarch64 over x86_64
+namespace = ["aarch64", "x86_64"]
+# prefer variants using hardware AES optimizations over these for a newer CPU
+# architecture
+feature.aarch64 = ["aes"]
+feature.x86_64 = ["aes"]
+# prefer x86-64-v3 and then older (even if the CPU is newer)
+property.x86_64.level = ["v3", "v2", "v1"]
+
+[variant.providers.aarch64]
+# Using different package based on the Python version
+requires = ["provider-variant-aarch64 >=0.0.1,<1"]
+# use only on aarch64/arm machines
+enable-if = "platform_machine == 'aarch64' or 'arm' in platform_machine"
+plugin-api = "provider_variant_aarch64.plugin:AArch64Plugin"
+
+[variant.providers.x86_64]
+requires = ["provider-variant-x86-64 >=0.0.1,<1"]
+# use only on x86_64 machines
+enable-if = "platform_machine == 'x86_64' or platform_machine == 'AMD64'"
+plugin-api = "provider_variant_x86_64.plugin:X8664Plugin"`
+```
+
+<figcaption>Listing 3. Example `pyproject.toml` with sort order defined</figcaption>
+</figure>
+
+However, no reason to stop at namespace. After all, a specific package may
+want to point out, say, that their AES-NI variant is more performant than
+a fallback implemented using the instruction sets provided by x86-64-v3.
+It might also be desirable reorder value; while
+I immediately can't think of a specific use case for that, let's have the option
+for symmetry anyway. So while the plugins specify the initial order of features
+and their values, packages are allowed to override it.
+
+On top of that, user configuration can be applied. After all, there is still
+a valid use case for the user to override the sort order defined by plugins
+and packages. And since it's no longer obligatory, having that is not a problem.
+
+## The null variant
+
+Let's consider that you are distributing a package with three CUDA variants,
+and a CPU variant.
+
+<figure>
+
+```rust
+torch-2.8.0-cp313-cp313-manylinux_2_28_x86_64-cu129.whl
```

Comment:
Previously in the article, the statement was that variants shouldn't be put in the tag, so this is somewhat surprising. Of course, ultimately different wheels need different file names. But what mechanism is supposed to be used for that hasn't been covered I think (or I missed it)